# SNP_explorer

SNP Explorer is a command-line tool for analyzing SNP variants from VCF (Variant Call Format) files. It filters variants based on clinical significance and gene of interest and generates detailed HTML and CSV reports.

## Features

1. Annotating variants using RSIDs or querying Ensembl to resolve missing RSIDs.
2. Querying dbSNP for clinical, frequency and gene annotations.
3. Filtering variants based on presence of RSID, clinical significance, and optional gene input.
4. Generating an HTML summary report and a CSV export of results.

## SNP_explorer Installation

```
git clone https://github.com/barbarapawlowskaa/SNP_explorer.git
```
## Dependencies Installation

```
cd SNP_explorer
python -m venv .venv
source .venv/bin/activate
pip install -r requirements.txt
```

## SNP_explorer Pipeline Usage

## Examples

